server/include/bge/World.h

```diff
@@ -14,22 +18,36 @@ namespace bge {
             void init();
             Entity createEntity();
 
-            template<typename ComponentType>
-            void addComponent(Entity e, ComponentType c);
+            // One function for each component type, since the alternatives involve crazy c++ that probably doesn't even work
+            void addComponent(Entity e, PositionComponent c);
+            void addComponent(Entity e, VelocityComponent c);
+            void addComponent(Entity e, MovementRequestComponent c);
+            void addComponent(Entity e, JumpInfoComponent c);
 
+            // No idea why we can do the simpler definition for deleteComponent but we can't for addComponent
             template<typename ComponentType>
             void deleteComponent(Entity e, ComponentType c);
 
-            void movePlayer(unsigned int player, float x, float y, float z);
+            void updateAllSystems();
+
+            // This can't be contained within a system since we want to do this as we receive client packets rather than once per tick
+            void updatePlayerInput(unsigned int player, glm::vec3 forward_direction, float pitch, float yaw, bool forwardRequested, bool backwardRequested, bool leftRequested, bool rightRequested, bool jumpRequested);
+
+            void fillInGameData(ServerToClientPacket& packet);
+
+            // void movePlayer(unsigned int player, float x, float y, float z);
 
             void printDebug();
 
         private:
-            std::vector<System> systems;
+            std::vector<std::unique_ptr<System>> systems;
             std::set<Entity> entities;
             int currMaxEntityId;
 
-            ComponentManager<PositionComponent> positionCM;
+            std::shared_ptr<ComponentManager<PositionComponent>> positionCM;
+            std::shared_ptr<ComponentManager<VelocityComponent>> velocityCM;
+            std::shared_ptr<ComponentManager<JumpInfoComponent>> jumpInfoCM;
+            std::shared_ptr<ComponentManager<MovementRequestComponent>> movementRequestCM;
```

Comment:
future: HealthComponent --- HealthManager (HP for players and Spring's dome shield etc.) 

damn, I'm starting to understand this ECS dialect